net/sync/gpoddernet/src/main/java/de/danoeh/antennapod/net/sync/nextcloud_gpodder/NextcloudGpodderService.java

```diff
@@ -0,0 +1,206 @@
+package de.danoeh.antennapod.net.sync.nextcloud_gpodder;
+
+import android.content.Context;
+import android.net.Uri;
+import android.os.Build;
+
+import androidx.annotation.RequiresApi;
+
+import com.google.gson.GsonBuilder;
+import com.nextcloud.android.sso.AccountImporter;
+import com.nextcloud.android.sso.aidl.NextcloudRequest;
+import com.nextcloud.android.sso.api.NextcloudAPI;
+import com.nextcloud.android.sso.exceptions.NextcloudFilesAppAccountNotFoundException;
+import com.nextcloud.android.sso.exceptions.NoCurrentAccountSelectedException;
+import com.nextcloud.android.sso.helper.SingleAccountHelper;
+import com.nextcloud.android.sso.model.SingleSignOnAccount;
+
+import org.json.JSONException;
+import org.json.JSONObject;
+
+import java.io.InputStream;
+import java.net.MalformedURLException;
+import java.net.URISyntaxException;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+
+import de.danoeh.antennapod.net.sync.gpoddernet.mapper.ResponseMapper;
+import de.danoeh.antennapod.net.sync.gpoddernet.model.GpodnetUploadChangesResponse;
+import de.danoeh.antennapod.net.sync.model.EpisodeAction;
+import de.danoeh.antennapod.net.sync.model.EpisodeActionChanges;
+import de.danoeh.antennapod.net.sync.model.ISyncService;
+import de.danoeh.antennapod.net.sync.model.SubscriptionChanges;
+import de.danoeh.antennapod.net.sync.model.SyncServiceException;
+import de.danoeh.antennapod.net.sync.model.UploadChangesResponse;
+
+import static java.time.Instant.now;
+
+public class NextcloudGpodderService implements ISyncService {
+
+    private final Context mContext;
+    private NextcloudAPI nextcloudAPI;
+
+    public NextcloudGpodderService(Context mContext) {
+        this.mContext = mContext;
+    }
+
+    @Override
+    public boolean isAuthenticated() {
+        try {
+            SingleAccountHelper.getCurrentSingleSignOnAccount(mContext);
+            return true;
+        } catch (NextcloudFilesAppAccountNotFoundException e) {
+            return false;
+        } catch (NoCurrentAccountSelectedException e) {
+            return false;
+        }
+    }
+
+    @Override
+    public void login() {
+        SingleSignOnAccount ssoAccount = null;
+        try {
+            try {
+                ssoAccount = AccountImporter.getSingleSignOnAccount(mContext, SingleAccountHelper.getCurrentSingleSignOnAccount(mContext).name);
+            } catch (NoCurrentAccountSelectedException e) {
+                e.printStackTrace();
+            }
+            NextcloudAPI.ApiConnectedListener callback = new NextcloudAPI.ApiConnectedListener() {
+                @Override
+                public void onConnected() {
```

Comment:
Do we have to wait for `onConnected` to be called before we are allowed to send requests?

Reply:
We dont have to wait for it to be called. Its part of the command chain until a request will be done.
I don't have the feeling i understand why you ask this.

Reply:
I feel like calling `nextcloudAPI.performNetworkRequest` before it is actually "ready" (eg. went through `onConnected`) could be a problem. At least that's how the Google Services connection works. For Google, it might take several seconds between creating the API client and it being actually ready to serve requests. (if I remember correctly, haven't worked with Google APIs in a while)

Reply:
i never had issues with unready connection